### PR TITLE
Media: avoid inserting .mov files from media picker

### LINKED ISSUE #4931: Video playback fails (.mov file) in unsupported browsers, try to avoid/fix
## Bug Description

This story was created using the WP plugin: https://www.tedisarah.com/web-stories/vegan-sneakers-2/

In page 3, there's a video that consistently shows "video failed to play" in Chrome/laptop. May also be Chrome-wide. 
image

It's a .mov video. **Only older versions of Safari, plus other browsers that supported Apple's QuickTime plugin.**

Can we explore solutions to this problem? Can we optimize/convert the video? If not, at a minimum we should discourage the use of .mov. 

## Additional Context

- Plugin Version:
- Operating System:
- Browser:

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### PR DESCRIPTION
**Pull request by embarks**
## Summary

- Took Pascal's suggestion to update the `MediaPane` to check for the allowed mimetypes. This filters it from the selection. See ticket below for the implementation suggestions.
- However, a user could still upload a `.mov` (though it would be filtered from the library) and then insert it from the picker. So, I throw an error and show a snackbar message like the drophandler does to prevent that. 

## Relevant Technical Choices
- I would have liked to prevent the upload of `.mov` files but the file blob comes in as a "image/jpeg" type from the `wp.media` upload module, so it bypasses the `isValidType` check that is already in place. Thus the error handling on the `MediaPane` itself.

## User-facing changes
When attempting to upload and then insert a `.mov` file the user should see the following message: 
Screen Shot 2020-11-03 at 1 45 55 PM

## Testing Instructions
- go to the local media tab on the media pane
- click upload
- upload a `.mov` file 
- The `.mov` file should NOT show in the gallery
- Attempt to insert the media by clicking Insert into Page
- Verify the media is not inserted and the snackbar message appears informing the user of the allowed media types. 

---

Fixes #4931

### CODE REVIEW

**Comment by zachhale** [comment]
This looks good to me, though the snackbar is very hard to see. When I tried it I didn't see the snackbar and was just confused why nothing happened. Is there a more prominent error mechanism that could be used?

**Comment by embarks** [comment]
> This looks good to me, though the snackbar is very hard to see. When I tried it I didn't see the snackbar and was just confused why nothing happened. Is there a more prominent error mechanism that could be used?

I'm using the snackbar to be consistent with the drop handler. If you drag/drop `.mov` it'll show the same thing... I'm not sure if there is a better mechanism. We could do a dismissable dialog?

**Comment by zachhale** [comment]
Okay, I'm not gonna block on it if it's consistent, just thought I'd point it out. I do think a dismissible dialog would be better UX.

**Review by spacedmonkey — CHANGES_REQUESTED** [review]
Please add an E2E test to stop future regression.

**Comment by swissspidy** [comment]
> Please add an E2E test to stop future regression.

E2E tests are quite tricky for videos because Chromium doesn't support many video codecs (e.g. no MP4). So I doubt that is possible.

File: assets/src/edit-story/components/library/panes/media/local/mediaPane.js [hunk 1]
```diff
@@ -130,16 +139,36 @@ function MediaPane(props) {
...
-      resource,
-      mediaPickerEl.sizes?.medium?.url || mediaPickerEl.url
-    );
+    try {
+      if (!allowedMimeTypes.includes(resource.mimeType)) {
+        /* translators: %s is a list of allowed file extensions. */
+        const message = sprintf(
+          /* translators: %s: list of allowed file types. */
+          __('Please choose only %s to insert into page.', 'web-stories'),
+          allowedMimeTypes.join(
```
Comment (swissspidy) [inline assets/src/edit-story/components/library/panes/media/local/mediaPane.js]: Use `allowedFileTypes` from `useConfig` for this instead.

Showing things like `image/png` to users is weird. `allowedFileTypes` gets you `png` etc.

File: assets/src/edit-story/components/library/panes/media/local/mediaPane.js [hunk 2]
```diff
@@ -130,16 +139,36 @@ function MediaPane(props) {
...
+      showSnackbar({
+        message: e.message,
+      });
+    }
   };
 
   const openMediaPicker = useMediaPicker({
     onSelect,
     onClose,
+    type: allowedMimeTypes,
```
Comment (swissspidy) [inline assets/src/edit-story/components/library/panes/media/local/mediaPane.js:172]: 👍

**Comment by spacedmonkey** [comment]
> E2E tests are quite tricky for videos because Chromium doesn't support many video codecs (e.g. no MP4). So I doubt that is possible.

Seems pretty simple to me. Add an example mov [file](https://github.com/WordPress/wordpress-develop/blob/master/tests/phpunit/data/uploads/small-video.mov?raw=true) on this [line](https://github.com/google/web-stories-wp/blob/main/bin/local-env/install-wordpress.sh#L92). 

Check to see you can't see that file does not exist in the media picker. Am I missing something?

**Comment by embarks** [comment]
getting this locally and in CI, not sure how to address... not happening on the other insert media e2e tests and I can't find role="checkbox" in the page... 
```
Expected page to pass Axe accessibility tests.
    Violations found:
    Rule: "landmark-one-main" (Document must have one main landmark)
    Help: https://dequeuniversity.com/rules/axe/3.5/landmark-one-main?application=axe-puppeteer
    Affected Nodes:
      html
        Fix ALL of the following:
          - Document does not have a main landmark.

    Rule: "list" (<ul> and <ol> must only directly contain <li>, <script> or <template> elements)
    Help: https://dequeuniversity.com/rules/axe/3.5/list?application=axe-puppeteer
    Affected Nodes:
      #__attachments-view-51
        Fix ALL of the following:
          - List element has direct children with a role that is not allowed: checkbox.

      194 |   }
      195 | 
    > 196 |   await expect(page).toPassAxeTests({
          |                      ^
      197 |     // Temporary disabled rules to enable initial integration.
      198 |     disabledRules: [
      199 |       'aria-allowed-role',
```

File: bin/local-env/install-wordpress.sh [hunk 3]
```diff
@@ -90,6 +90,7 @@ echo -e $(status_message "Import default set of media assets...")
 wp media import /var/www/html/wp-content/e2e-assets/example-1.jpg
 wp media import /var/www/html/wp-content/e2e-assets/example-2.jpg
 wp media import /var/www/html/wp-content/e2e-assets/example-3.png
+wp media import /var/www/html/wp-content/e2e-assets/small-video.mov
```
Comment (spacedmonkey) [inline bin/local-env/install-wordpress.sh:93]: 😄

File: tests/e2e/specs/editor/media/insertMovVideo.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    await page.evaluate((selector) => {
+      document.querySelector(selector).click();
+    }, btnTab);
+    const btnSelector =
+      '.attachments-browser .attachments .attachment:first-of-type';
+    await page.waitForSelector(btnSelector);
+    await expect(page).not.toMatchElement('.type-video.subtype-quicktime');
+    await percySnapshot(page, 'Avoid inserting .mov files');
+
+    const closeBtnSelector = '.media-modal-close';
```
Comment (spacedmonkey) [inline tests/e2e/specs/editor/media/insertMovVideo.js:50]: Why do you need these lines?

File: bin/local-env/install-wordpress.sh (hunk 3 above)
Comment (spacedmonkey) [inline bin/local-env/install-wordpress.sh:93]:
```suggestion
wp media import /var/www/html/wp-content/e2e-assets/small-video.mov
wp media import /var/www/html/wp-content/e2e-assets/small-video.mp4
```

Is it worth adding an MP4 here, so we can see that both jpg and mp4 are both working as expected?

**Comment by spacedmonkey** [comment]
This is great that we now have a test for this!

Had one possible improvement ( I think ) and a quick question. Otherwise this is looking good 👍

File: tests/e2e/specs/editor/media/insertMovVideo.js (hunk 4 above)
Comment (embarks) [reply tests/e2e/specs/editor/media/insertMovVideo.js:50]: the ones after the snapshot? I shouldn't have deleted my last comment :D If I don't close the media picker dialog after the snapshot the e2e tests fail with this error:
```
  ● Inserting .mov from dialog › should not list the .mov

    expect(received).toPassAxeTests(expected)

    Expected page to pass Axe accessibility tests.
    Violations found:
    Rule: "landmark-one-main" (Document must have one main landmark)
    Help: https://dequeuniversity.com/rules/axe/3.5/landmark-one-main?application=axe-puppeteer
    Affected Nodes:
      html
        Fix ALL of the following:
          - Document does not have a main landmark.
    
    Rule: "list" (<ul> and <ol> must only directly contain <li>, <script> or <template> elements)
    Help: https://dequeuniversity.com/rules/axe/3.5/list?application=axe-puppeteer
    Affected Nodes:
      #__attachments-view-51
        Fix ALL of the following:
          - List element has direct children with a role that is not allowed: checkbox.
    

      194 |   }
      195 | 
    > 196 |   await expect(page).toPassAxeTests({
          |                      ^
      197 |     // Temporary disabled rules to enable initial integration.
      198 |     disabledRules: [
      199 |       'aria-allowed-role',

      at runAxeTestsForStoriesEditor (config/bootstrap.js:196:22)
      at Object.<anonymous> (config/bootstrap.js:232:3)
```
I don't think I control this rendering (the thumbnails with checkboxes inside the media picker) so I simply closed the dialog.

File: bin/local-env/install-wordpress.sh (hunk 3 above)
Comment (embarks) [reply bin/local-env/install-wordpress.sh:93]: I think that's probably worth it, it's an easy test. :)

**Review by BrittanyIRL — APPROVED** [review]
works as expected for me 👍

File: bin/local-env/install-wordpress.sh (hunk 3 above)
Comment (spacedmonkey) [reply bin/local-env/install-wordpress.sh:93]: Let's add this then.

**Review by spacedmonkey — APPROVED** [review]
👍

**Comment by embarks** [comment]
Just adding a quick e2e test for the mp4 👍

**Comment by embarks** [comment]
The e2e tests are more fragile than expected... trying to make minimal changes so that the other insert media e2e tests pass :)

**Comment by embarks** [comment]
@spacedmonkey we might need to make a new bug instead... :( I'm going to remove the extra mp4 testing for now, but we should write that e2e test for the follow up.

I uploaded the MP4 of the small sample video and inserted it into the page from the upload media library... although the source shows that it is available in the "resource" in my React dev tools, the preview claims there was no "src" and the video fails to play:

Screen Shot 2020-11-05 at 12 21 57 PM
Screen Shot 2020-11-05 at 12 22 30 PM
Screen Shot 2020-11-05 at 12 22 53 PM

I'm not sure why, because the display element works fine and uses the same props to show and play the video. It works fine in the editor, the video only fails to play in Preview and Publish modes.

Edit: it failed with another uploaded MP4. There seems to be a problem with the output rendering of local mp4s.

**Comment by embarks** [comment]
created the bug here: #5142

**Pull request merged**